### PR TITLE
Allow "this" in datums

### PR DESCRIPTION
**Pull request by awgrover**
It seems more convenient to write "this" in a datum, rather than the node's own name when referring to another datum. Especially if the node changes name. 

I suspect that I have the "caller" wrong in:
void Proxy::setThis(PyObject* proxy, Root* for_this, Downstream* caller)
{
    // "caller" probably wrong
    boost::python::extract<Proxy*>(proxy)()->for_this = Proxy::makeProxyFor(for_this, caller);
}

### CODE REVIEW

**Comment by mkeeter** [comment]
Thanks for this contribution!

I like the idea, but I think there's a cleaner way to implement it.

How about modifying `Graph::pyGetAttr`?
```cpp
PyObject* Graph::pyGetAttr(std::string name, Downstream* caller,
                           uint8_t flags) const
{
    ...
    if (name == "this")
        if (auto d = dynamic_cast<Datum*>(caller))
            return Proxy::makeProxyFor(d->parentNode(), caller, flags)
    ...
}
```

I also suspect that this will break change tracking: `this` isn't actually a node name, so when a node informs the graph that it's changed, whatever looked it up as `this` won't necessarily re-evaluate.

I'd like to see a unit test that confirms it works.

**Comment by astronouth7303** [comment]
I would prefer it be called `self`, since that's the conventional Python name.

@mkeeter How does the change tracker work? Lexically or by what's referenced? Does it matter, since it's only referring to the node that's being recalculated?

**Comment by awgrover** [comment]
Whatever is cleaner. My first attempt at understanding how the symbol lookup worked. I swear I looked at that method. "self", of course, is more python.

Ah, tracking. Ok, I will try to figure out a test (and use your patch). May take me days.

File: lib/graph/tests/datum.cpp [hunk 1]
```diff
@@ -269,3 +269,44 @@ TEST_CASE("Datum output detection")
...
+      REQUIRE(PyFloat_AsDouble(fromself->currentValue()) == 123.0);
+      REQUIRE(PyFloat_AsDouble(expr->currentValue()) == 124.0);
+      // REQUIRE(PyFloat_AsDouble(showfromself->currentValue()) == 123.0);
+    }
+
+    SECTION("change datum")
+    {
+    x->setText("231.0");
+    REQUIRE(PyFloat_AsDouble(fromself->currentValue()) == 231.0);
+    REQUIRE(PyFloat_AsDouble(expr->currentValue()) == 232.0);
```
Comment (awgrover) [inline lib/graph/tests/datum.cpp]: These were intended to test "tracking". The basic "change value" seems to work fine (cf. mkeeter sub-graph lookup by __parent: is that a similar case?). What other tests would show that "tracking" is/isn't working?

File: lib/graph/tests/datum.cpp [hunk 2]
```diff
@@ -269,3 +269,44 @@ TEST_CASE("Datum output detection")
...
 }
+
+TEST_CASE("'self'")
+{
+    auto g = new Graph();
+    auto n = new Node("a", g);
+    auto x = new Datum("x", "123.0", &PyFloat_Type, n);
+    auto fromself = new Datum("fromself", "self.x", &PyFloat_Type, n);
+    auto expr = new Datum("expr", "self.x + 1", &PyFloat_Type, n);
+    // auto showfromself = new Datum("showfromself", Datum::SIGIL_OUTPUT + "fromself", &PyFloat_Type, n);
```
Comment (awgrover) [inline lib/graph/tests/datum.cpp]: I'd like to test that "output" datums get their value from a datum that uses "self". I don't know how to construct a datum for that.

**Comment by mkeeter** [comment]
I didn't notice the recent changes until just now -- I'll check them out this weekend.

**Comment by mkeeter** [comment]
This is now merged (after a tiny bit of rebasing / cleaning).

Thanks for the PR, @awgrover — this is the first outside contribution to the graph engine!

**Pull request closed without merging**